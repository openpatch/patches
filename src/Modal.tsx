/** @jsx jsx */
import { ReactNode, MouseEvent, KeyboardEvent } from "react";
import { jsx, css, Theme } from "@emotion/react";
import ReactModal, { Props } from "react-modal";
import { ClassNames } from "@emotion/react";
import { darken, transparentize } from "polished";

import { Card } from "./Card";
import { useBreakpoints, useTheme } from "./hooks";

export type ModalProps = {
  /**
   * Determines if the modal is visible or not.
   */
  open?: boolean;
  onClose?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * React Modal's accessibility string.
   */
  contentLabel?: string;
  children?: ReactNode;
  /**
   * The element that should be used as root for the
   * React portal used to display the modal. See
   * http://reactcommunity.org/react-modal/accessibility/#app-element
   */
  appElement?: string | HTMLElement;
} & Partial<Props>;

export const TRANSITION_DURATION = 200;
const TOP_MARGIN = "10vh";
const TRANSFORM_Y_FLOATING = "10vh";
const FLOATING_TRANSITION = `${TRANSITION_DURATION}ms ease-in-out`;
const FIXED_TRANSITION = `${TRANSITION_DURATION}ms cubic-bezier(0, 0.37, 0.64, 1)`;

const makeQuery = (breakpoint: string) => {
  return `@media screen and (min-width: ${breakpoint})`;
};

/**
 * Wrapper component for ReactModal. Uses the Card component
 * to wrap content passed as the children prop. Don't forget to set
 * the aria prop when using this.
 * http://reactcommunity.org/react-modal/accessibility/#aria
 */
export const Modal = ({
  open = true,
  onClose = () => {},
  contentLabel = "Modal",
  appElement = "#root",
  children,
  ...props
}: ModalProps) => {
  const [theme] = useTheme();

  // in testing there is not appElement. Therefore, this would throw an error.
  if (process.env.NODE_ENV !== "test") {
    ReactModal.setAppElement(appElement);
  }

  return (
    <ClassNames>
      {({ css }) => {
        // React Modal styles
        // https://reactcommunity.org/react-modal/styles/classes/

        const className = {
          base: css`
            label: modal;
            outline: none;
            bottom: 0;
            max-height: 80vh;
            -webkit-overflow-scrolling: touch;
            overflow-y: auto;
            position: fixed;
            transform: translateY(100%);
            transition: transform ${FIXED_TRANSITION};
            width: 100%;
            width: 100vw;

            ${makeQuery(theme.breakpoints[0].toString())} {
              transition: transform ${FLOATING_TRANSITION},
                opacity ${FLOATING_TRANSITION};
              margin: ${TOP_MARGIN} auto auto;
              max-height: 90vh;
              max-width: 90%;
              min-width: 450px;
              opacity: 0;
              position: relative;
              transform: translateY(${TRANSFORM_Y_FLOATING});
            }
            ${makeQuery(theme.breakpoints[1].toString())} {
              max-width: 720px;
            }
          `,
          afterOpen: css`
            label: modal--after-open;
            transform: translateY(0);

            ${makeQuery(theme.breakpoints[0].toString())} {
              opacity: 1;
              transform: translateY(0);
            }
          `,
          beforeClose: css`
            label: modal--before-close;
            transform: translateY(100%);
            ${makeQuery(theme.breakpoints[0].toString())} {
              opacity: 0;
              transform: translateY(${TRANSFORM_Y_FLOATING});
            }
          `,
        };

        const overlayClassName = {
          base: css`
            label: modal__overlay;
            background: ${transparentize(
              0.8,
              darken(0.8, theme.colors.background)
            )};
            bottom: 0;
            left: 0;
            opacity: 0;
            position: fixed;
            right: 0;
            top: 0;
            transition: opacity 200ms ease-in-out;
            z-index: ${theme.zIndices[50]};
            ${makeQuery(theme.breakpoints[0].toString())} {
              -webkit-overflow-scrolling: touch;
              overflow-y: auto;
            }
          `,
          afterOpen: css`
            label: modal__overlay--after-open;
            opacity: 1;
          `,
          beforeClose: css`
            label: modal__overlay--before-close;
            opacity: 0;
          `,
        };

        const reactModalProps = {
          isOpen: open,
          className,
          overlayClassName,
          htmlOpenClassName: "ReactModal__Html--open",
          shouldCloseOnOverlayClick: true,
          contentLabel,
          onRequestClose: onClose,
          closeTimeoutMS: TRANSITION_DURATION,
          ...props,
        };
        return (
          <ReactModal {...reactModalProps}>
            <Card boxShadow="xlarge">{children}</Card>
          </ReactModal>
        );
      }}
    </ClassNames>
  );
};
