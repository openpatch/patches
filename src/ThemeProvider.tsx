import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
