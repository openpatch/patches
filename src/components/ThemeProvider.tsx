import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
