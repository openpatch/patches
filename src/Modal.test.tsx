import React from "react";
import { render } from "@testing-library/react";
import ReactModal from "react-modal";
import { Modal, ModalProps } from "./Modal";
import * as MockedModal from "./Modal";
import { ThemeProvider } from "./ThemeProvider";

ReactModal.setAppElement(document.createElement("div"));

const renderWithTheme = (props?: ModalProps) =>
  render(
    <ThemeProvider>
      <Modal {...props}>Hi</Modal>
    </ThemeProvider>
  );

beforeEach(() => {
  (MockedModal as any).TRANSITION_DURATION = 0;
});

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot when closed", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});

test("should match snapshot when open", () => {
  const { container } = renderWithTheme({ open: true });
  expect(container).toMatchSnapshot();
});
