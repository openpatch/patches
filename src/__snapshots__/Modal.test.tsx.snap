// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when closed 1`] = `
<div>
  <div
    id="patches-root"
  />
</div>
`;

exports[`should match snapshot when open 1`] = `
<div>
  <div
    id="patches-root"
  />
</div>
`;
